### ISSUE TITLE
Optimize performance 

### ISSUE THREAD

**Issue by balajtimate** [zavolanlab/htsinfer#109]
Currently, HTSinfer with `--records` set to 0 (default) runs for hours per sample.

Run test script with `--records` set to 200k, 500k, 1 million, 0 (all records).

**Comment by uniqueg** [comment]
The point here is that we should analyze how many reads we need to process to come to a conclusion. There are two general approaches here, a simple one, and a more tricky one:

1. Analyze lots of samples and find out the minimum number of records required after which, say, 95% of samples yield the exact same results - for all the inferred metadata - that is achieved when _all_ records are processed. We would then set that number of records as default.
2. We introduce logic that, for each inference functionality, consumes records continually and evaluates whether we have reached a reliable and definite outcome. This has the advantage that for every sample only the minimum required number of records is consumed (rather than one default for _all _ samples) and that, similarly, for each functionality only the minimum number of records is consumed. However, this is difficult for several reasons: (1) we likely need to rewrite very large chunks of the code; (2) it may not be easy to tell when to stop without having probabilistic models for each inference category that would tell us when we have enough confidence.

As option 2. is currently not feasible, this issue should focus on two things:
1. Determining a good one-size-fits-all (well, not all, but maybe 95%) default for the number of records
2. Analyze which parts of the code take how long, so that we have an idea what we could try to optimize. The more detailed this analysis, the better (i.e., it's one thing to know that, e.g., organism inference takes longer than 3' adapter inference, but it would be even better if we also knew how long index generation and mapping take)

**Comment by balajtimate** [comment]
To get a better picture for the general runtime, I ran the tests with the `--records` parameter set to 1 million, 5 million, 10 million and 0 (default). Unfortunately, the default run timed out after 3 days, and I only got the results for 63 samples, so I only used the first 63 sample results from the other runs as well. I broke down the whole HTSinfer run into 5 processes to which seemed to differ most between runs:
 - Processing of reads: reading in the records and writing them to temp
 - Extracting read length: parsing the reads and determining the length
 - Quantification using Kallisto: using the quantified reads to determine the library source, there's also an indexing step but that took ~10 seconds regardless of the records setting
 - Alignment of reads with STAR: to determine read orientation
 - Cutadapt: to get read layout and adapter sequence

Here are the results, the duration of each process (wall-clock time) converted to seconds:
![perf_test_results](https://github.com/zavolanlab/htsinfer/assets/51365402/37347ac0-0b04-4182-b4ef-08761fabe1be)

As for the results of the inference, there was no difference between the 1mil, 5mil and 10mil sets in any of the metadata inferred. There was 9 samples were these results (predicted organism, adapter, orientation, read length min, max) differed from the default set, but 4 out of those 9 were due to an error. So the predictions were different in 5 cases:
- The orientation in 2 samples: predicted unstranded (default records run) and SF (1-5-10mil run) in one sample, and vice-versa in another one
- Read length minimum in 2 samples: in one case for the default it was predicted as 63 instead of 64 (1-5-10mil), and in another 35 instead of 59. 
- 1 sample: no adapter was inferred for the 0 records run, while it was identified for the others.

And as I mentioned, there were 4 samples, were the 0 records run ran to an error, in 2 cases it was due to Cutadapt (I'll look into this and open a separate issue), and 2 had the `FileProblem: [Errno 28] No space left on device`, which is kinda weird, but I think the reason for that is that temp had no space left, also one of them was a PE sample with 183318939 reads in each fastq, and that might have been too much.

I know that the reduced number of samples (63) is not ideal, I could rerun the 0 records with the 770 samples, though it would take quite long. Based on this, I would say we could set the default to 10 million records, or since there's no difference between the results with 1 million, we could go lower and speed up the whole thing.

**Comment by uniqueg** [comment]
Really cool, thanks a lot!

How do you run the samples? Do you do it from one cluster job? Or did 770-63 jobs time out? If you ran them from one cluster job, I guess you could optimize this by starting several parallel runs (up to 770, but perhaps you can pool them such that you have enough slots in your quota to run them all in parallel, maybe ~40 jobs with 20 samples each). It might also help with the TEMP space, maybe (if there is a quota per job, I don't know).

About lowering the default: I would first look what happens if you have the results for all (or at least most) samples when run with all reads (`--records=0`). If I interprete your analysis correctly, there were 5 discrepancies in 59 comparisons (not counting the errors). That's almost 10%. And given that the 1 million run records are so fast (looks like ~2-3 min), I don't think it's worth going much lower, though you are welcome to try 500k and 100k, even if it's just to find the limit of when the accuracy starts to deteriorate.

Good job also on recording the runtimes separately for the different processes. This will help to optimize the speed if we need to. Btw, it looks to me that the times, both for the entire runs and the individual processes, are roughly linear. If true, that's great coding :) Perhaps you could actually check this. For example by calculating the Pearson correlation coefficients for each sample (runtime vs number of processed records) and then plotting a histogram of the distributions of the coefficients. In addition, you could also create a a density plot of runtimes vs processed records for all samples. And maybe you find a way to calculate the slope if you indeed find that the runtimes scale roughly linear with the number of processed samples.

**Comment by balajtimate** [comment]
> How do you run the samples? Do you do it from one cluster job? Or did 770-63 jobs time out? If you ran them from one cluster job, I guess you could optimize this by starting several parallel runs (up to 770, but perhaps you can pool them such that you have enough slots in your quota to run them all in parallel, maybe ~40 jobs with 20 samples each). It might also help with the TEMP space, maybe (if there is a quota per job, I don't know).

Yeah, it was one job, and the QoS was set to three days (I assumed that wouldn't be enough for all 770 samples but I didn't think it would be this low 😅). So far it hasn't been an issue, but I'll run them again in parallel.

> And given that the 1 million run records are so fast (looks like ~2-3 min), I don't think it's worth going much lower, though you are welcome to try 500k and 100k, even if it's just to find the limit of when the accuracy starts to deteriorate.

I also wouldn't go lower, I meant setting 1 million as the default. I wouldn't expect to be much difference between 1mil and 500k (don't know about 100k though), but sure, I'll run it and see what happens. And when I have the runtimes from all the different record setting, I'll also check for linearity.

**Comment by uniqueg** [comment]
Yeah, parallel runs at least for `--records=0`, but also for the other ones should speeed up things a lot in terms of the return time. If you optimize things a bit (I guess you can try up to 100 jobs), maybe you can get results for all samples within an hour for anything up to 10M records, and maybe 8h or so for full-sized samples.

**Comment by balajtimate** [comment]
I ran the test again with different `--records`, with parallel runs, this time on all of 770 samples. Sorry it took this long, but it took multiple tries, because after 10 million, there is an increase in the number of errors with the tmp directory in `/scratch` ('No space left on device' or 'Creation of temporary directory failed'). With the records set to 0 (def), almost 1/3 of samples (254 of 770) failed because of this, even when running only 5 samples per job. 
![perf_test_results](https://github.com/zavolanlab/htsinfer/assets/51365402/f355c1d4-5bc9-434d-94ff-892e9fabe0e6)
So the runtimes are much longer on average (compared to last time, when it was only 63 samples), but the same linearity seems to be there between increasing records and runtime (up until 50 million, with the increase in errors, it kind of breaks down).

I'm working on the plot that shows how the predicted _results_ (org, adapter, length) change between runs with different records, I'll post it as soon as I have that along with the distributions of the Pearson-correlations.

**Comment by uniqueg** [comment]
Good stuff - thanks a lot! Looking forward to the correlation with results :)

**Comment by balajtimate** [comment]
So here's some of the results from the runs with different records - sorry that it's just these barplots again, I made heatmaps showing the different results between the runs, but it turned out really unreadable/uninformative. 
![0708_htsinfer_performance_results](https://github.com/zavolanlab/htsinfer/assets/51365402/ed1b9c2d-459d-406f-a496-0b9e345ac0cb)
Most samples didn't have an orientation and adapter from the metadata, so there couldn't be a match between the given orientation/adapter and the predicted ones, that's why there's a large percentage of FALSE (and Undecided is when there were no orient/adapter predicted at all).

**Comment by uniqueg** [comment]
Thanks. I'm not sure I understand that last sentence. If there cannot be a true prediction (because we don't know the real orientation and/or adapter), how can there be a _false_ prediction then?

I also _really_ don't understand the orientation and adapter results for 100 million reads. I would assume that the vast majority of libraries have less than or at least not considerably more than 100 million reads. How could the results be so much worse than for 50 million or when considering _all_ reads? I really think there is some kind of problem. As a starter: Could you please plot a distribution of the library sizes?

Also, do you have any ideas why the predictions are getting rather worse for large sample sizes, especially for the organism and read length estimates?

Also, in principle, the read length couldn't even be wrong when looking at _all_ reads. It could just be that the reported number is wrong, or what is reported does not match well with the min/max or mean read length (which one are you using for the comparison) - for whatever reason. It would be good to look deeper here into what is happening. If outliers (i.e., a few reads that are longer or less long than the expected/reported read length), then it's probably best to use the median and/or the mode of the read length.

Some cosmetic comments:
- It would be really helpful to put a percent scale on the axes on the right hand sides.
- It would make it easier to interpret these plots if the colors were used consistently, i.e., the same colors were used for `Undecided`, `FALSE` and `TRUE` across all plots.

**Comment by balajtimate** [comment]
I updated the plots based on what we discussed and what you mentioned here:
![1708_htsinfer_performance_results](https://github.com/zavolanlab/htsinfer/assets/51365402/ad63d0f8-6788-48f5-8851-6e9e11adacbc)
I corrected the error with the 100 million run. Undecided is if nothing was predicted, Match when the predicted value matched what was given in the metadata, and Mismatch otherwise - even if the metadata is unknown, at least we can see from the predicted adapters, that around 60% of samples get something predicted. 

Here is the distribution of the library sizes:
![1708_htsinfer_read_counts](https://github.com/zavolanlab/htsinfer/assets/51365402/66d25731-8632-425f-a896-ea393e9ae2de)
So the majority of libraries is indeed below 100 million, I'm guessing we would still see a marginal increase in quality in 50-100-0 runs compared to the results of 10 million; I'm trying to fix the scratch storage issue now.

**Comment by uniqueg** [comment]
Great, thanks a lot. Would still be good to separate mismatch for cases where there is no ground truth from mismatch for cases where the predicted result does not match the ground truth

**Comment by balajtimate** [comment]
Okay, so I reran the whole analysis again with a more streamlined approach on the cluster (setting the threads for STAR to run on and only keeping the result files) that seems to have solved the problem of no space left on tmp. I also separated the categories, added "No_metadata" when there was no prior information. 
![0811_htsinfer_performance_results](https://github.com/zavolanlab/htsinfer/assets/51365402/3a8e848c-7c95-45dc-b06c-154aa2045314)
There's still some fluctuation, due to some issue with Cutadapt, when using a large number of records, but the 50mil-100mil-0def (default, using all records) is fixed at least. 
And here are the computing times in seconds:
![perf_test_results](https://github.com/zavolanlab/htsinfer/assets/51365402/4706e1e2-bd04-44ae-b339-a9a36dcc5234)
So based on this, should we set a default for the records to use param or not? I was initially leaning towards using 1mil (because of the lowest number of org mismatches) or 10mil (because of a reasonable time for the whole run), but even with the (current) default 0, the avg time for the total run is ~1.5 hours, which is not that much I think, and given that the majority of the samples are <50mil, I guess it wouldn't really make sense to not use all the records by default?
